src/main/protobuf/remote_output_service.proto

```diff
@@ -0,0 +1,305 @@
+import "google/protobuf/empty.proto";
+
+option java_package = "com.google.devtools.build.lib.remote";
+option java_outer_classname = "RemoteOutputServiceProto";
+option go_package = "remoteoutputservice";
+
+// The Remote Output Service may be used by users of the Remote
+// Execution API to construct a directory on the local system that
+// contains all output files of a build.
+//
+// Primitive implementations of this API may simply download files from
+// the Content Addressable Storage (CAS) and store them at their
+// designated location. Complex implementations may use a pseudo file
+// system (e.g., FUSE) to support deduplication, lazy loading and
+// snapshotting.
+service RemoteOutputService {
+  // Methods that can be invoked at any point in time.
+
+  // Clean all data associated with a single output path, so that the
+  // next invocation of StartBuild() yields an empty output path. This
+  // may be implemented in a way that's faster than removing all of the
+  // files from the file system manually.
+  rpc Clean(CleanRequest) returns (google.protobuf.Empty);
+
+  // Signal that a new build is about to start.
+  //
+  // The client uses this call to obtain a directory where outputs of
+  // the build may be stored, called the output path. Based on the
+  // parameters provided, the remote output service may provide an empty
+  // output path, or one that has contents from a previous build of the
+  // same project.
+  //
+  // In case the output path contains data from a previous build, the
+  // remote output service is responsible for calling
+  // ContentAddressableStorage.FindMissingBlobs() for all of the objects
+  // that are stored remotely. This ensures that these objects don't
+  // disappear from the Content Addressable Storage while the build is
+  // running. Any files that are absent must be removed from the output
+  // path and reported through InitialOutputPathContents.modified_paths.
+  rpc StartBuild(StartBuildRequest) returns (StartBuildResponse);
+
+  // Methods that can only be invoked during a build.
+
+  // Create one or more files, directories or symbolic links in the
+  // output path.
+  rpc BatchCreate(BatchCreateRequest) returns (google.protobuf.Empty);
+
+  // Obtain the status of one or more files, directories or symbolic
+  // links that are stored in the input path.
+  rpc BatchStat(BatchStatRequest) returns (BatchStatResponse);
+
+  // Signal that a build has been completed.
+  rpc FinalizeBuild(FinalizeBuildRequest) returns (google.protobuf.Empty);
+}
+
+message CleanRequest {
+  // The output base identifier that was provided to
+  // StartBuildRequest.output_base_id whose data needs to be removed.
+  string output_base_id = 1;
+}
+
+message StartBuildRequest {
+  // A client-chosen value that uniquely identifies this output path.
+  // This value must be set to ensure that the remote output service is
+  // capable of managing builds concurrently.
```

Comment:
ROS could manage builds concurrently even without it, based on build_id, right? I guess this is mostly optimization - so that subsequent builds for same output_base_id can be incremental?

Reply:
Let me respond to this question, while also responding to a part of [this comment](https://github.com/bazelbuild/bazel/pull/12823#discussion_r588793838):

> it could allow ROS to perform concurrent builds in a single workspace - not applicable for bazel, but maybe other clients, or perform clean if ROS stores just previous build and one from request does not match.

When designing this protocol, I tried to make it so that implementations of the Remote Output Service may differ in terms of smartness. bb_clientd uses FUSE, so it could do all sorts of fancy virtualization tricks (snapshotting, copy-on-write, etc.). But it should still be possible to build a dumb daemon that does nothing more than offload the 'bottom half' of Bazel. With that I mean: it should be possible to develop a gRPC service that just manages a concrete (native) directory on your file system. The dumbest versions of such a daemon just download files when BatchCreate() is called, while others may use OS specific features to create placeholder files that 'trap' read requests, downloading the files on demand.

In situations like these, it isn't realistic to assume that you can do efficient snapshotting/cloning of build directories. It likely means you need to do recursive copying, which I don't think is acceptable. This is one of the reasons the `output_base_id` exists, and that at most one active build for an `output_base_id` may exist.

Eliminating `output_base_id` entirely also means that the Remote Output Service doesn't really know to which workspace build directories belong. It would make it hard (impossible?) to implement Clean(), right?

Reply:
I've included parts of the phrasing that you proposed. I did not include the change to let output_base_id be "[a-zA-Z0-9]{1,64}", as that allows the use of names (e.g., AUX, COM1, LPT3) that Windows does not permit.